src/clj/fluree/db/query/fql/parse.cljc

```diff
@@ -436,7 +436,10 @@
 (defmethod parse-pattern :union
   [[_ & unions] vars context]
   (let [parsed (mapv (fn [clause]
-                       (parse-where-clause clause vars context))
+                       (if (and (vector? clause)
```

Comment:
Why couldn't this be built into `parse-where-clause` instead of special casing here? Could we have the same issues with other higher level pattern types like `optional` that operate on nested where clauses?

The only constraint I can think of with this approach is that where clauses consisting of only a single non-map pattern is required to be wrapped in a vector, but I think, if this is the reason, then it should be handled in the sparql translator and not here.

Reply:
The problem is that our `::where` grammar is ambiguous:
```
[:union
  [:query1 ...]
  [:query2 ...]]
```
Could either be two single-pattern clauses or one multi-pattern clause. All of our higher-order patterns indicate that they accept multiple clauses, but looking at the parser we assume for all but `:union` that we are only receiving a single clause.

I think for the short term I'll generalize the condition so that it will work on all higher-order patterns by changing it to this:
```
(and (vector? clause) (keyword? (first clause))) ;; clause is a single pattern
```
But I think we'll need what our actual grammar is - is the parser the source of truth or is the malli schema?

Reply:
Ok, I think I found a better way to handle it. `parse-where-clause` was calling `util/sequential` to ensure that regardless of whether the clause was a single where-pattern or a sequence of where-patterns we would treat it as a sequence. However, all non-node where patterns are already sequential, so this was not wrapping it in a vector. Now we correctly handle all single-pattern types and turn them into a sequence of where-patterns.

Reply:
Union is supposed to handle multiple clauses. A clause can either be a map or a vector. A map is the case of a singleton subject-node map (with possible nested maps in the case where the value of a property is a reference). 

If it is a vector, then it is a vector of where patterns. A single vector is not valid. 

**Single-pattern where clauses need to be wrapped in vectors unless they are node maps**, and this is what [the malli schema](https://github.com/fluree/db/blob/main/src/clj/fluree/db/validation.cljc#L323-L325) should be checking for. 

```
[:union
  [:query1 ...]
  [:query2 ...]]
```
is not valid. It should be:
```
[:union
  [[:query1 ...]]
  [[:query2 ...]]]
```

Reply:
Ok, there's the source of the confusion. I got the test syntax from @bplatz , who also seemed confused that we only supported single node patterns instead of single where patterns of any type.

So, now that we have a way to support single where patterns of any type, do we want that or no?

Reply:
> A single vector is not valid
This is not clear. I meant to say, a single vector pattern is not valid unless it is wrapped in another vector. so you can't have the following inside of a where clause
```
{...
 :where [:optional {:foo/bar ?baz}]
 ...}
```
It must be:
```
{...
 :where [[:optional {:foo/bar ?baz}]]
 ...}
```

The rules for what can come in an optional and union are the same as what can come in a where clause.

Reply:
> Ok, I think I found a better way to handle it. `parse-where-clause` was calling `util/sequential` to ensure that regardless of whether the clause was a single where-pattern or a sequence of where-patterns we would treat it as a sequence. However, all non-node where patterns are already sequential, so this was not wrapping it in a vector. Now we correctly handle all single-pattern types and turn them into a sequence of where-patterns.

I'm not sure why this would have ever happened, unless there was a bug in validation. If it got past the validation step, then any single where patterns that were sequential should have already been wrapped. It must have been that the `:query` pattern type wasn't being validated properly.

Reply:
> Ok, there's the source of the confusion. I got the test syntax from @bplatz , who also seemed confused that we only supported single node patterns instead of single where patterns of any type.
> 
> So, now that we have a way to support single where patterns of any type, do we want that or no?

I had considered checking for a keyword as the first element when I first wrote the query parser but then I decided against it. I just can't remember _why_ I decided against it. 

Since I can't remember, then it probably wasn't that good of a reason not to do it, so let's go with it for now and we can revisit if I remember why I didn't do it in the first place and/or some problems pop up.